cmd/mtaBuild.go

```diff
@@ -0,0 +1,390 @@
+	if err != nil {
+		log.Entry().
+			WithError(err).
+			Fatal("failed to execute mta build")
+	}
+}
+
+func runMtaBuild(config mtaBuildOptions,
+	commonPipelineEnvironment *mtaBuildCommonPipelineEnvironment,
+	e envExecRunner,
+	p piperutils.FileUtils,
+	httpClient piperhttp.Sender) error {
+
+	e.Stdout(log.Entry().Writer()) // not sure if using the logging framework here is a suitable approach. We handover already log formatted
+	e.Stderr(log.Entry().Writer()) // entries to a logging framwork again. But this is considered to be some kind of project standard.
+
+	var err error
+
+	handleSettingsFiles(config, p, httpClient)
+
+	handleDefaultNpmRegistry(config, e)
+
+	mtaYamlFile := "mta.yaml"
+	mtaYamlFileExists, err := p.FileExists(mtaYamlFile)
+
+	if err != nil {
+		return err
+	}
+
+	if !mtaYamlFileExists {
+
+		if err = createMtaYamlFile(mtaYamlFile, config.ApplicationName, p); err != nil {
+			return err
+		}
+
+	} else {
+		log.Entry().Infof("\"%s\" file found in project sources", mtaYamlFile)
+	}
+
+	if err = setTimeStamp(mtaYamlFile, p); err != nil {
+		return err
+	}
+
+	mtarName, err := getMtarName(config, mtaYamlFile, p)
+
+	if err != nil {
+		return err
+	}
+
+	var call []string
+
+	switch config.MtaBuildTool {
+
+	case "classic":
+
+		mtaJar := getMarJarName(config)
+
+		buildTarget, err := ValueOfBuildTarget(config.BuildTarget)
+
+		if err != nil {
+			return err
+		}
+
+		call = append(call, "java", "-jar", mtaJar, "--mtar", mtarName, fmt.Sprintf("--build-target=%s", buildTarget), "build")
+		if len(config.Extensions) != 0 {
+			call = append(call, fmt.Sprintf("--extension=%s", config.Extensions))
+		}
+
+	case "cloudMbt":
+
+		platform, err := ValueOfBuildTarget(config.Platform)
+		if err != nil {
+			return err
+		}
+
+		call = append(call, "mbt", "build", "--mtar", mtarName, "--platform", platform.String())
+		if len(config.Extensions) != 0 {
+			call = append(call, fmt.Sprintf("--extensions=%s", config.Extensions))
+		}
+		call = append(call, "--target", "./")
+
+	default:
+
+		return fmt.Errorf("Unknown mta build tool: \"%s\"", config.MtaBuildTool)
+	}
+
+	if err = addNpmBinToPath(e); err != nil {
+		return err
+	}
+
+	log.Entry().Infof("Executing mta build call: \"%s\"", strings.Join(call, " "))
+
+	if err := e.RunExecutable(call[0], call[1:]...); err != nil {
+		return err
+	}
+
+	commonPipelineEnvironment.mtarFilePath = mtarName
+	return nil
+}
+
+func getMarJarName(config mtaBuildOptions) string {
+
+	mtaJar := "mta.jar"
+
+	if len(config.MtaJarLocation) > 0 {
+		mtaJar = config.MtaJarLocation
+	}
+
+	return mtaJar
+}
+
+func addNpmBinToPath(e envExecRunner) error {
+	path := "./node_modules/.bin"
+	oldPath := os.Getenv("PATH")
+	if len(oldPath) > 0 {
+		path = path + ":" + oldPath
+	}
+	e.Env(append(os.Environ(), "PATH="+path))
+	return nil
+}
+
+func getMtarName(config mtaBuildOptions, mtaYamlFile string, p piperutils.FileUtils) (string, error) {
```

Comment:
For the nexus upload, I would need a similar function. Could we make this re-usable (i.e. move it to pkg and export it)?

Reply:
... you need the same strategy? Yes, in this case we can of course refactor accordingly.

Reply:
Well, I'm not sure if I need the same strategy, but I don't want to have my own version of the truth. Such things should be unified and commoditized in the whole application imo.

Reply:
as already mentioned above, we can refactoring that out. For the sake of an better overview: to my knowledge we have currently these approaches:

1.) in the `mtaBuild` step we calculate the name of the mtar file. Here I see no other option --> we need some name
2.) in the `neoDeploy` we rely on `commonPipelineEnvironment.mtarFilePath`. That property is set by `mtaBuild`. To my understanding basically the same mechanism like in groovy is also available in the go approach, hence we can share also here.
3.) in `cloudFoundryDeploy` we simply traverse the file system with pattern `**/*.mta` and hope to find only one entry.

--> what I would like to say: there are several ways to get that file name. Before starting refactoring we should carefully discuss which approach we should choose.

Reply:
One solution I see is that the location of the .mta is returned from a function. That function should live in the pkg tree, for example as "mtautils.go". Then it is used from all three locations. What currently exists in the mtaBuild implementation to calculate the path should be the implementation of that function.

Reply:
I think we should not share state between steps, but should implement behaviour in a consistent way, so the way `mtaBuild` determines which file to produce should also be available to other steps.